File: straw/media.py
import re
from typing import Dict, List, Any
from bs4 import BeautifulSoup
from .client import StrawClient

class MediaScraper:
    def __init__(self, **client_options):
        self.client = StrawClient(**client_options)

    async def extract_media(self, url: str) -> Dict[str, Any]:
        html = await self.client.get_text(url)
        soup = BeautifulSoup(html, 'lxml')

        page_title = soup.title.string.strip() if soup.title and soup.title.string else ""
        media_links = set()

        for tag in soup.find_all(['video', 'audio', 'source', 'img']):
            src = tag.get('src') or tag.get('srcset')
            if src:
                urls = re.findall(r'''https?:\/\/[^\s"',]+''', src)
                for u in urls:
                    media_links.add(u)
                if src.startswith('http') and src not in media_links:
                    media_links.add(src)

        for tag in soup.find_all('a', href=True):
            href = tag.get('href')
            if href and href.startswith('http') and re.search(r'\.(pdf|doc|docx|xls|xlsx|ppt|pptx|txt|csv|rtf|mp4|mp3|webm|wav|ogg|m4a|avi|mkv|mov|flv|png|jpg|jpeg|gif|svg|webp|avif|ico|bmp)(?:\?.*)?$', href, re.IGNORECASE):
                media_links.add(href)

        raw_links = re.findall(r'''https?:\/\/[^\s"',]+\.(?:png|jpg|jpeg|gif|svg|webp|avif|ico|bmp|pdf|doc|docx|xls|xlsx|ppt|pptx|txt|csv|rtf|mp4|mp3|webm|wav|ogg|m4a|avi|mkv|mov|flv)''', html, re.IGNORECASE)
        for link in raw_links:
            media_links.add(link)

        return {
            'pageTitle': page_title,
            'mediaLinks': list(media_links)
        }

File: src/scrapers/media.ts
import * as cheerio from 'cheerio';
import { StrawClient, StrawClientOptions } from '../core/client';

export interface MediaScrapeResult {
    pageTitle: string;
    mediaLinks: string[];
}

export class MediaScraper {
    private client: StrawClient;

    constructor(options?: StrawClientOptions) {
        this.client = new StrawClient(options);
    }

    /**
     * Attempts to find direct media files (Images/Audio/Video/Documents) referenced in any generic webpage HTML.
     */
    public async extractMedia(url: string): Promise<MediaScrapeResult> {
        const html = await this.client.getText(url);
        const $ = cheerio.load(html);

        const pageTitle = $('title').text().trim();
        const mediaLinks = new Set<string>();

        // 1. Check <video>, <audio>, <img>, and <source> tags
        $('video, audio, img, source').each((_, el) => {
            const src = $(el).attr('src') || $(el).attr('srcset');
            if (src) {
                // handle srcset parsing simply by grabbing the first URL if needed, or just finding http links
                const urls = src.match(/https?:\/\/[^\s"',]+/g);
                if (urls) urls.forEach(u => mediaLinks.add(u));
                else if (src.startsWith('http')) mediaLinks.add(src);
            }
        });

        // 2. Check <a> tags for document/media links
        $('a').each((_, el) => {
            const href = $(el).attr('href');
            if (href && href.startsWith('http') && href.match(/\.(pdf|doc|docx|xls|xlsx|ppt|pptx|txt|csv|rtf|mp4|mp3|webm|wav|ogg|m4a|avi|mkv|mov|flv|png|jpg|jpeg|gif|svg|webp|avif|ico|bmp)(\?.*)?$/i)) {
                mediaLinks.add(href);
            }
        });

        // 3. Fallback: Check regex for embedded JSON or JS containing media/document links
        const rawLinksMatch = html.match(/https?:\/\/[^\s"',]+\.(png|jpg|jpeg|gif|svg|webp|avif|ico|bmp|pdf|doc|docx|xls|xlsx|ppt|pptx|txt|csv|rtf|mp4|mp3|webm|wav|ogg|m4a|avi|mkv|mov|flv)/gi);
        if (rawLinksMatch) {
            for (const link of rawLinksMatch) {
                mediaLinks.add(link);
            }
        }

        return {
            pageTitle,
            mediaLinks: Array.from(mediaLinks)
        };
    }
}
